### PR TITLE
Login

### PR DESCRIPTION
**Pull request by shrutiyer**
ADD: Login page and Registration page. Also, logged in middle-ware to redirect if user isn't logged in.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -5,9 +5,13 @@ var logger = require('morgan');
 var cookieParser = require('cookie-parser');
 var bodyParser = require('body-parser');
 var mongoose = require('mongoose');
+var passport = require('passport');
+var LocalStrategy = require('passport-local').Strategy;
 
 var index = require('./routes/index');
 var users = require('./routes/users');
+// requires the model with Passport-Email plugged in
```
Comment (davidabrahams) [inline app.js]: fix comment

File: app.js [hunk 2]
```diff
@@ -19,7 +23,7 @@ app.set('view engine', 'hbs');
 //app.use(favicon(path.join(__dirname, 'public', 'favicon.ico')));
 app.use(logger('dev'));
 app.use(bodyParser.json());
-app.use(bodyParser.urlencoded({ extended: false }));
```
Comment (davidabrahams) [inline app.js:22]: @shrutiyer we should check if this actually matters

File: oauth.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// This is where we would put environment ID's, just in case we wanted to sign
```
Comment (davidabrahams) [inline oauth.js]: oops didn't mean to commit this file

File: routes/index.js [hunk 4]
```diff
@@ -2,28 +2,47 @@ var express = require('express');
...
+    passport.authenticate('local')(req, res, function () {
+      res.send({redirect: '/'});
+    });
+  });
 }
 
 routes.GETindex = function(req, res){
-	// Student.find
-	// res.JSON({students: allStudents});
+  res.sendFile('index.html', { root: path.join(__dirname, '../views') });
```
Comment (davidabrahams) [inline routes/index.js]: did we accidentally overwrite something important?

File: app.js (hunk 2 above)
Comment (shrutiyer) [inline app.js:22]: Maybe it doesn't. I'll check it tonight

Comment (shrutiyer) [inline app.js:22]: Just checked. It doesn't matter. Resetted it to false

**Comment by shrutiyer** [comment]
Made changes to the pull req according to @greneroom comments

File: app.js [hunk 5]
```diff
@@ -29,13 +33,25 @@ app.use(require('node-sass-middleware')({
 }));
 app.use(express.static(path.join(__dirname, 'public')));
 
-// app.use('/', routes);
+app.use(require('express-session')({
+  secret: 'Brenna is Bae',
```
Comment (nmohamed) [inline app.js]: export this

File: app.js [hunk 6]
```diff
@@ -82,5 +105,17 @@ app.use(function(err, req, res, next) {
   });
 });
 
+function isLoggedIn(req, res, next) {
+
+  // if user is authenticated in the session, carry on
+  if (req.isAuthenticated())
+  {
```
Comment (nmohamed) [inline app.js]: move bracket up, could get rid of empty lines 109/115

File: public/javascripts/apps.js [hunk 7]
```diff
@@ -0,0 +1 @@
+var loginApp = angular.module('loginApp', []);
```
Comment (nmohamed) [inline public/javascripts/apps.js:1]: point of this file?

**Comment by aconsilvio** [comment]
:+1:

File: public/javascripts/loginCtrl.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  $scope.register = function() {
+    console.log($scope.email);
+    var data = {username: $scope.email, password: $scope.password};
+    console.log(data)
+    $http.post("register", data).then(function (response) {
+      console.log('success!');
+      console.log(response.data);
+      window.location.href = response.data.redirect;
+    }, function (response) {
+      console.log('error!');
```
Comment (nmohamed) [inline public/javascripts/loginCtrl.js]: put this and 18/19 into one console.log

File: views/login.html [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    <!-- The above 3 meta tags *must* come first in the head; any other head content must come *after* these tags -->
+    <meta name="description" content="">
+    <meta name="author" content="">
+
+    <title>Spring Initiative Login</title>
+    <!-- Latest compiled and minified CSS -->
+    <link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap.min.css">
+
+    <!-- jQuery library -->
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.0/jquery.min.js"></script>
```
Comment (nmohamed) [inline views/login.html]: if we're not using jquery could get rid of since it'll improve loading a tiny bit

File: app.js (hunk 5 above)
Comment (davidabrahams) [inline app.js]: what do you mean?

File: public/javascripts/apps.js (hunk 7 above)
Comment (davidabrahams) [inline public/javascripts/apps.js:1]: I figured this would be where we declared all the Angular apps.

File: views/login.html [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+          <label>
+            <input type="checkbox" value="remember-me"> Remember me
+          </label>
+        </div>
+        <button class="btn btn-lg btn-primary btn-block" type="submit">Sign in</button>
+        <button type='button' class="btn btn-lg btn-primary btn-block" ng-click="register()">Register</button>
+
+      </form>
+    </div> <!-- /container -->
+    <script type="text/javascript" src="/javascripts/apps.js"></script>
```
Comment (nmohamed) [inline views/login.html:43]: put these with other scripts

File: models/studentDataModel.js [hunk 11]
```diff
@@ -16,13 +16,7 @@ var studentSchema = mongoose.Schema({
...
 });
 
-var userSchema = mongoose.Schema({
-	username: String,
-	password: String
-});
-
 module.exports.attendace = mongoose.model('Attendace', dataSchema);
 module.exports.grades = mongoose.model('Grades', dataSchema);
 module.exports.data = mongoose.model('Data', dataSchema);
```
Comment (nmohamed) [inline models/studentDataModel.js]: not really part of this commit, but lets separate the data and student model into two separate files, since they dont rely on each other anymore and dont need to be together

File: routes/index.js [hunk 12]
```diff
@@ -2,28 +2,49 @@ var express = require('express');
...
-  res.sendFile('index.html', { root: path.join(__dirname, '../views') });
-}
-
 routes.POSTlogin = function(req, res, next) {
-  res.sendFile('index.html', { root: path.join(__dirname, '../views') });
+  passport.authenticate('local', function(err, user, info) {
+    if (err) { return res.status(500).send(err.message); }
+    if (!user) {
+      return res.status(401)
+      .send(info.message);}
```
Comment (nmohamed) [inline routes/index.js]: bracket

File: routes/index.js [hunk 13]
```diff
@@ -2,28 +2,49 @@ var express = require('express');
...
 
 routes.GETindex = function(req, res){
-	// Student.find
-	// res.JSON({students: allStudents});
+  // Student.find
+  // res.JSON({students: allStudents});
+  res.sendFile('index.html', { root: path.join(__dirname, '../views') });
+}
+
+routes.GETlogin = function(req, res){
```
Comment (nmohamed) [inline routes/index.js]: delete GETlogin - i dont think we still need it right

File: routes/index.js [hunk 14]
```diff
@@ -2,28 +2,49 @@ var express = require('express');
...
+    passport.authenticate('local')(req, res, function () {
+      res.send({redirect: '/'});
+    });
+  });
 }
 
 routes.GETindex = function(req, res){
-	// Student.find
-	// res.JSON({students: allStudents});
+  // Student.find
```
Comment (nmohamed) [inline routes/index.js:41]: should add something in GETindex to redirect you to `/login` if you aren't logged in

File: views/login.html [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    <div class="container">
+      <form role="form" class="form-signin" ng-submit="login()">
+        <h2 class="form-signin-heading">Please sign in</h2>
+        <label for="inputEmail" class="sr-only">Email address</label>
+        <input type="email" id="inputEmail" class="form-control" placeholder="Email address" ng-model="email" autofocus>
+        <label for="inputPassword" class="sr-only">Password</label>
+        <input type="password" id="inputPassword" class="form-control" placeholder="Password" ng-model="password">
+        <div class="checkbox">
+          <label>
+            <input type="checkbox" value="remember-me"> Remember me
```
Comment (nmohamed) [inline views/login.html:35]: this doesnt work yet right?

File: views/login.html [hunk 16]
```diff
@@ -0,0 +1,47 @@
+
```
Comment (nmohamed) [inline views/login.html:1]: question about angular: is there a way to have the equivalent of main.html so every page doesnt require you to have stuff outside the `<body>` tags, and login.html would just be the inner parts?

File: app.js (hunk 5 above)
Comment (nmohamed) [inline app.js]: i'm assuming we don't want people to know what this is, so this would mean making it an environment variable (like the mongoURI)

**Comment by nmohamed** [comment]
:dragon: looks fierce :dragon_face:

File: routes/index.js (hunk 14 above)
Comment (davidabrahams) [inline routes/index.js:41]: this is already happening in `app.js`

File: views/login.html (hunk 15 above)
Comment (davidabrahams) [inline views/login.html:35]: correct. I'll add a github issue

File: views/login.html (hunk 10 above)
Comment (davidabrahams) [inline views/login.html:43]: it's supposed to be here. The script in `head` load before the page loads. You want angular to inject after the page loads. Placing it at the end of the `body` ensures this happens.

File: views/login.html (hunk 16 above)
Comment (davidabrahams) [inline views/login.html:1]: If the entire page functioned on one page, then yes. However, since we're right now sending multiple HTML files, you need to define the header on each one. When we have more function HTML working later, we will refactor it to work on one page, and therefore only have one header.

**Comment by davidabrahams** [comment]
Awaiting merge confirmation from @nmohamed and @BrennaManning

**Comment by davidabrahams** [comment]
Oops closed it on accident. That's my bad.

**Comment by nmohamed** [comment]
all good :pager:

**Comment by davidabrahams** [comment]
Okay. We need Brennas approval then we're good to merge

**Comment by BrennaManning** [comment]
Okay, it looks good! I'm going to merge it.
